File: javascript/cote/LTCD/역순_연결_리스트2.mjs

const TEST_CASES = 10
const N = 100

class Node {
	constructor(value, next = null) {
		this.value = value
		this.next = next
	}

	*[Symbol.iterator]() {
		let cur = this

		while (cur) {
			yield cur.value
			cur = cur.next
		}

	}
}

function initialize() {
	const n = Math.floor(Math.random() * (N - 2)) + 2

	const root = new Node(0)
	let node = root

	for (let i = 0; i < n; i++) {
		node.next = new Node(node.value + 1)
		node = node.next
	}

	const end = Math.floor(Math.random() * (n - 1)) + 1
	const start = Math.floor(Math.random() * end)

	return [root, start, end]
}

function solve(node, start, end) {
	const root = new Node(-1, node)
	let [prev, cur] = [root, root.next]

	for (let i = 0; i < start; i++) {
		[prev, cur] = [prev.next, cur.next]
	}

	// prev ->  ... -> cur -> next -> next.next
	// prev -> next-> ... -> cur -> next.next
	for (let i = 0; i < end - start; i++) {
		let tmp = prev.next
		{	
			[prev.next, cur.next] = [cur.next, cur.next.next]
			prev.next.next = tmp
		}

	}

	return root.next
}

function main() {
	for (let i = 0; i < TEST_CASES; i++) {
		const [node, start, end] = initialize()
		console.log(...solve(node, start, end))
	}
}

main()

File: python/cote/LTCD/역순_연결_리스트2.py
from __future__ import annotations
from typing import TypeVar, Generic
import random

TEST_CASES = 10
N = 100

T = TypeVar("T")


class Node(Generic[T]):

    def __init__(self, value: T, next: Node[T] = None):
        self.value = value
        self.next = next

    def __iter__(self):
        node = self

        while node:
            yield node.value
            node = node.next


def solve(node: Node[T], n, m):

    root = start = Node(None, node)

    for _ in range(n - 1):
        start = start.next

    end = start.next

    for _ in range(m - n):
        tmp, start.next, end.next = start.next, end.next, end.next.next
        start.next.next = tmp

    return root.next


if __name__ == "__main__":
    random.seed(43)

    for _ in range(TEST_CASES):
        n = random.randrange(1, N)
        end = random.randrange(1, n)
        start = random.randrange(1, end)

        cur = n
        prev = Node(cur)
        cur -= 1

        while cur > 0:
            prev = Node(cur, prev)
            cur -= 1
        print(start, end)
        print([*solve(prev, start, end)])
